src/incremental/serialize.ml

```diff
@@ -43,49 +40,78 @@ let results_exist () =
   let r_str = Fpath.to_string r in
   Sys.file_exists r_str && Sys.is_directory r_str
 
-(* Convenience enumeration of the different data types we store for incremental analysis, so file-name logic is concentrated in one place *)
-type incremental_data_kind = SolverData | CilFile | VersionData | AnalysisData
-
-let type_to_file_name = function
-  | SolverData -> solver_data_file_name
-  | CilFile -> cil_file_name
-  | VersionData -> version_map_filename
-  | AnalysisData -> analysis_data_file_name
-
-(** Used by the server mode to avoid serializing the solver state to the filesystem *)
-let server_solver_data : Obj.t option ref = ref None
-let server_analysis_data : Obj.t option ref = ref None
-
-(** Loads data for incremental runs from the appropriate file *)
-let load_data (data_type: incremental_data_kind) =
-  if server () then
-    match data_type with
-    | SolverData -> !server_solver_data |> Option.get |> Obj.obj
-    | AnalysisData -> !server_analysis_data |> Option.get |> Obj.obj
-    | _ -> failwith "Can only load solver and analysis data"
-  else
-    let p = Fpath.(gob_results_dir Load / type_to_file_name data_type) in
-    unmarshal p
-
-(** Stores data for future incremental runs at the appropriate file, given the data and what kind of data it is. *)
-let store_data (data : 'a) (data_type : incremental_data_kind) =
-  if server () then
-    match data_type with
-    | SolverData -> server_solver_data := Some (Obj.repr data)
-    | AnalysisData -> server_analysis_data := Some (Obj.repr data)
-    | _ -> ()
-  else (
+(** Module to cache the data for incremental analaysis during a run, before it is stored to disk, as well as the server mode *)
+module Cache = struct
+  type t = {
+    solver_data: Obj.t option ref;
+    analysis_data: Obj.t option ref;
+    version_data: MaxIdUtil.max_ids option ref;
+    cil_file: Cil.file option ref;
```

Comment:
Instead of `ref`s, the fields could simply be `mutable`.

Also, why are they all `option`s? When would we ever have incremental data that doesn't have all four components present? I think we always need all of them (or have none of them), so the whole record could be optional.

Reply:
>Instead of refs, the fields could simply be mutable.

Done.

>Also, why are they all options? When would we ever have incremental data that doesn't have all four components present? I think we always need all of them (or have none of them), so the whole record could be optional.

The issue is that during the initial analysis, these fields would not all become available at the same time. Namely, the Cil file and the max ids would become available after parsing (in `Maingoblint`), while the solver and analysis data only becomes availabe after solving the constraint system. 

If one were to make the fields in the record non-optional, one would have to drag the cil file and max ids around until all of the data is available, so that one can set them in the record here.

Reply:
> these fields would not all become available at the same time

Right. At some point this bothered me before, but I think we can keep them optional for now.

At some point it might be useful to have them all in an immutable record that's read and created as a whole, because storing them at different times maybe can cause some inconsistencies during server mode or such. For example, if the new file is parsed and stored, the solving starts but is aborted, then the cache possibly has inconsistent state.